### PR TITLE
fix(profiling): use service lock to avoid race condition on start

### PR DESCRIPTION
**Pull request by jd**
While unlikely, there's a possible race condition when starting a service
multiple times with multiple parallel threads.

### CODE REVIEW

File: ddtrace/profiling/_periodic.py [hunk 1]
```diff
@@ -141,16 +142,17 @@ def run(self):
                     raise
 
 
-def PeriodicRealThread(*args, **kwargs):
+def PeriodicRealThreadClass():
+    # type: () -> typing.Type[PeriodicThread]
     """Create a PeriodicRealThread based on the underlying thread implementation (native, gevent, etc).
```
Comment (P403n1x87) [inline ddtrace/profiling/_periodic.py:147]: This is no longer creating an object, but returning a class

File: ddtrace/profiling/_periodic.py [hunk 2]
```diff
@@ -141,16 +142,17 @@ def run(self):
                     raise
 
 
-def PeriodicRealThread(*args, **kwargs):
+def PeriodicRealThreadClass():
```
Comment (P403n1x87) [inline ddtrace/profiling/_periodic.py:145]: A capitalised function name could be misleading, but if this formatting is already adopted elsewhere I'm fine with it

Comment (P403n1x87) [reply ddtrace/profiling/_periodic.py:145]: Perhaps one could do
~~~ python
class PeriodicRealThread():
    def __new__(cls):
        ...

Comment (jd) [reply ddtrace/profiling/_periodic.py:145]: I think it's fine as it's meant to be returning a `class`.

File: ddtrace/profiling/profiler.py [hunk 3]
```diff
@@ -276,9 +276,9 @@ def copy(self):
             service=self.service, env=self.env, version=self.version, tracer=self.tracer, tags=self.tags
         )
 
-    def start(self):
+    def _start(self):
+        # type: () -> None
         """Start the profiler."""
-        super(_ProfilerInstance, self).start()
```
Comment (Kyle-Verhoog) [inline ddtrace/profiling/profiler.py:281]: Seems like the profiler `start` was already thread-safe because of this call?

**Review by Kyle-Verhoog — DISMISSED** [review]
Seems like `ProfilerInstance` was already thread-safe but the change still makes sense and this creates a safer pattern to ensure all Service instances will be thread-safe to start :+1:

File: ddtrace/profiling/profiler.py (hunk 3 above)
Comment (jd) [reply ddtrace/profiling/profiler.py:281]: No, the lock is released by the time `super` returns 😢 
In practice I don't think anyone ever tried to start a `Profiler` in two different threads, but hey that's still a bug if we want a `Service` to be thread-safe 😉

Comment (Kyle-Verhoog) [reply ddtrace/profiling/profiler.py:281]: Oh right of course the race was actually in `profiler.start` and not the super! 👍

**Review by Kyle-Verhoog — APPROVED** [review]

**Review by P403n1x87 — APPROVED** [review]

**Pull request merged**